apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs

```diff
@@ -401,22 +410,52 @@ Task<IEnumerable<long>> ISpannerTransaction.ExecuteBatchDmlAsync(ExecuteBatchDml
         public new Task<DateTime> CommitAsync(CancellationToken cancellationToken = default)
         {
             GaxPreconditions.CheckState(Mode != TransactionMode.ReadOnly, "You cannot commit a readonly transaction.");
-            var request = new CommitRequest { Mutations = { _mutations }, ReturnCommitStats = LogCommitStats, RequestOptions = BuildCommitRequestOptions() };
-            return ExecuteHelper.WithErrorTranslationAndProfiling(async () =>
+            // Return the original response if this transaction has already successfully committed.
+            // Committing the same transaction twice on Cloud Spanner will return the same commit
+            // timestamp, and there's no need for us to repeat the RPC here when we already know the
+            // result. This also allows us to more aggressively return the session to the pool after
+            // committing/rolling back a transaction.
+            if (_hasCommittedOrRolledBack && _commitResponse != null)
             {
-                var callSettings = SpannerConnection.CreateCallSettings(settings => settings.CommitSettings, CommitTimeout, cancellationToken);
-                var response = await _session.CommitAsync(request, callSettings).ConfigureAwait(false);
-                if (response.CommitTimestamp == null)
-                {
-                    throw new SpannerException(ErrorCode.Internal, "Commit succeeded, but returned a response with no commit timestamp");
-                }
-                if (LogCommitStats)
+                return Task.FromResult(_commitResponse.CommitTimestamp.ToDateTime());
+            }
+            CheckNotHasCommittedOrRolledBack();
+            _hasCommittedOrRolledBack = true;
```

Comment:
> The only thing that we are blocking with this would be if someone has added manual retries of actual transient errors, instead of relying on the standard retry mechanism built into the library.

This and users who continue to retry transient errors past the standard retry mechanism. Both would constitute breaking changes, so if we were to this we'd need to include it in a new major version. Note that this is not the only breaking change this PR introduces, see my notes on the previous review.

But even if we were to accept the breaking change, as a user, I would still find it unexpected to have my transaction rolledback in the presence of transient errors, even if those were already retried some by the library.
I think it would be fine to rollback in the presence of non-transient errors like the `ALREADY_EXIST`  example you mentioned earlier. Although even this would be a breaking change.

In general I agree with the aim to release the session as soon as possible. But I also think that we should be cautious and continue to support some reasonable use cases like users wanting to retry further some transient errors. We really haven't gotten any reports stemming from sessions not beeing released soon enough, so I'd rather we don't go to the other extreme.

Reply:
> This and users who continue to retry transient errors past the standard retry mechanism. Both would constitute breaking changes, so if we were to this we'd need to include it in a new major version. Note that this is not the only breaking change this PR introduces, see my notes on the previous review.

Yes, I agree that this is a breaking change.

> ...continue to support some reasonable use cases like users wanting to retry further some transient errors.

I would prefer to keep this as simple as possible, and accept that we don't support that. Currently, the client library itself does not determine whether an error is transient or not. Instead, we have the following layering:
1. If a gRPC invocation fails, Gax determines whether an error is transient and retryable. If it is, and the retry settings allow it to be retried (e.g. the deadline has not been exceeded), then Gax retries it. This is all transparent to the client library.
2. Any error that escapes 1, either because Gax did not consider it retryable or because the retry settings disallowed it to be retried further, is bubbled up to the client application.

The only error that the client library handles specifically in some cases are `Aborted` errors. These cause the entire transaction to be retried when a RetryableTransaction is used.

If we were to support that the user can retry transient errors after Gax has given up, then we need some way to determine what constitutes a transient error in the client library, and add that as an extra layer between the two mentioned above. Do we consider all error codes that have been registered as retryable for the `Commit` RPC retryable? E.g. should a `DEADLINE_EXCEEDED` error be considered transient in this case if it is registered as a retryable error code, and Gax has given up because the total timeout has been exceeded?

> We really haven't gotten any reports stemming from sessions not beeing released soon enough

There's suspicion that this was a contributing factor in a case where the application got itself completely stuck. (Not the error handling, but the late returnal of sessions to the pool in general)

Reply:
I still prefer not to rollback at all on failure, even on any failure, given that Spanner itself does not rollback failed transactions. But I do see your points. If we go with "we rollback on any failure approach" we need to document that very clearly, not just because it's currently a breaking change, but also because in my opinion, it'll be very unexpected behaviour for users.

Also, note that even if we don't rollback (and release) failed transactions, with this change, succesfull or explicitly rollbacked transactions (hopefully way more common than failed transactions) will still be released sooner. My point, it's likely that the already scarced issues on "sessions not being released soon enough" should be greatly improved even if we don't rollback failed transactions.

Also, should Spanner consider automatically rolling back unrecoverabled transactions and send a signal back to clients.

@jskeet for your thoughts on this

Reply:
> Also, should Spanner consider automatically rolling back unrecoverabled transactions and send a signal back to clients.

Spanner actually does that (and my example with an `ALREADY_EXISTS` error was a bad one in this case, as Spanner actually considers that an unrecoverable error). So simple errors, like for example syntax errors, allow you to continue with the transaction as long as you catch them in application code. Other errors, like a constraint violation for an insert statement, are treated by Spanner as uncrecoverable errors. Trying to commit the transaction after catching such an error in application code will still fail with the same error.

Spanner does however not include any information in the error that indicates whether it is an unrecoverable error or not. You will only know that if you still try to continue with the transaction.

Reply:
Honestly, that makes me think that the appropiate thing would be for Spanner to signal back to the client which transactions have been rolled back. For those we can release the session back to the pool and not for the others. It's the backend who can make that decision safely I think.

Given that, regardless, this is breaking change (because of the dispose behaviour aspect), so we'd want to include it on the next major version that won't happen inmediately, do you think Spanner will consider sending that kind of signal back to clients?

Reply:
I'd definitely like more information from the Spanner server.

One aspect to drill down on:

> There's suspicion that this was a contributing factor in a case where the application got itself completely stuck. (Not the error handling, but the late returnal of sessions to the pool in general)

I don't remember the details, but if that was due to a customer bug (e.g. not disposing of SpannerTransaction or similar) then I'm less convinced it's needed - at least for that reason. Typically, papering over the symptoms of a customer bug just means some other symptom for the same bug comes up later.

That doesn't mean it's not worth doing - but I wouldn't want "it masks one impact of a customer bug" to be a justification.